haystack/utils/import_utils.py

```diff
@@ -31,25 +31,32 @@ def safe_import(import_path: str, classname: str, dep_group: str):
     try:
         module = importlib.import_module(import_path)
         classs = vars(module).get(classname)
+        if classs is None:
+            raise ImportError(f"Failed to import '{classname}' from '{import_path}'")
     except ImportError as ie:
         classs = _missing_dependency_stub_factory(classname, dep_group, ie)
     return classs
 
 
+class MissingDependency:
+    pass
+
+
 def _missing_dependency_stub_factory(classname: str, dep_group: str, import_error: Exception):
     """
     Create custom versions of MissingDependency using the given parameters.
     See `safe_import()`
     """
 
-    class MissingDependency:
+    class MissingOptionalDependency(MissingDependency):
```

Comment:
I think we should leave the name as `MissingDependency` or update to `MissingOptionalDependency`. I don't think it is necessary to have the empty `MissingDependency` class above.

Reply:
I used MissingDependency in past commits, but then I removed to comply with mypy because the way the project structure is set up.

The point of this dummy-class is that we use `safe_import` for the default initialization of optional nodes. 
It returns an object which can be any of the nodes, or `MissingOptionalDependency`, which uses a nested class.
Preparing for an improvement in project typing support, I decided to create this dummy-class to allow type checking in many places (despise not implementing in whole package). It was a preparation.

If desired, I can remove and in future we open another PR to improve all places where the type checking fails because of this.